# DSCC 201/401 Final Project -  Question1

EXTRA CREDIT 1 (10 points): Use the Keras library as provided by the Python 3 (miniforge 23.3.1-1 dsc-2024) kernel to build a simple artificial neural network for the prediction of the solar radio flux based on the sunspot number and the individual number of X-ray and optical solar flares. Begin with importing your cleaned data file (solar_data_clean.csv) into a Panadas data frame. Partition the data set appropriately and train your neural network. What is the predicted radio flux for a day when 73 sunspots are observed on the sun and 2 solar flares with C-class energy levels and 1 solar flare with an M-class energy level occur? How confident are you in the answer Explain your reasoning. Please provide a PDF of your Jupyter notebook showing ALL input and output and upload BOTH the PDF and the Jupyter notebook file (ExtraCredit1.ipynb and ExtraCredit1.pdf) showing all input and output.

In [5]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

# Load the cleaned data file
df = pd.read_csv("solar_data_clean.csv")
df.head()

,Date,Radio_Flux_10.7_cm,SESC_Sunspot_Number,Sunspot_Area,New_Regions,GOES15_XRay_Bkgd_Flux,XRay_C,XRay_M,XRay_X,S,Optical_1,Optical_2,Optical_3
0,1997-01-01,72,0,0,0,A0.5,0,0,0,1,0,0,0
1,1997-01-02,72,0,0,0,A0.5,0,0,0,1,0,0,0
2,1997-01-03,73,0,0,0,A0.5,0,0,0,1,0,0,0
3,1997-01-04,74,13,10,1,A0.6,0,0,0,1,0,0,0
4,1997-01-05,74,15,20,0,A0.7,0,0,0,1,0,0,0


In [6]:
# Select feature and target columns
feature_cols = [
    "SESC_Sunspot_Number",
    "XRay_C", 
    "XRay_M", 
    "XRay_X",
    "S",
    "Optical_1", 
    "Optical_2", 
    "Optical_3"
]

target_col = "Radio_Flux_10.7_cm"

X = df[feature_cols].to_numpy()
y = df[target_col].to_numpy()

# Partition the data set into training set and test set
# standardize the features
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Build a neural network model
n_features = X_train_scaled.shape[1]

model = keras.Sequential([
    layers.Input(shape=(n_features,)),
    layers.Dense(32, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1)
])

model.compile(
    optimizer="adam",
    loss="mse",
    metrics=["mae"]
)

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 32)             │           288 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 833 (3.25 KB)

 Trainable params: 833 (3.25 KB)

 Non-trainable params: 0 (0.00 B)

In [7]:
# Train the neural network model
model.fit(
    X_train_scaled, y_train,
    validation_data=(X_test_scaled, y_test),
    epochs=100,
    batch_size=32,
    verbose=1
)

Epoch 1/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 13593.4980 - mae: 107.9848 - val_loss: 3775.2136 - val_mae: 53.0760
Epoch 2/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 2205.4619 - mae: 36.0957 - val_loss: 722.1401 - val_mae: 18.8043
Epoch 3/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 653.0989 - mae: 17.8400 - val_loss: 517.8265 - val_mae: 15.4959
Epoch 4/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 512.2719 - mae: 15.1984 - val_loss: 416.5073 - val_mae: 13.6283
Epoch 5/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 431.1604 - mae: 13.5991 - val_loss: 354.3467 - val_mae: 12.7184
Epoch 6/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 372.7403 - mae: 12.6185 - val_loss: 310.7935 - val_mae: 11.7431
Epoch 7/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 321.0659 - mae: 11.8151 - val_loss: 288.4671 - val_mae: 11.3131
Epoch 8/100
264/264 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 311.4010 - mae: 11.6424 - val_loss: 273.4472 - val_mae: 11.

In [8]:
# Evaluate the Model Performance on the test set
test_mse, test_mae = model.evaluate(X_test_scaled, y_test, verbose=0)
print("Test MSE:", test_mse)
print("Test MAE:", test_mae)

Test MSE: 213.64012145996094
Test MAE: 9.953743934631348


In [22]:
# Use the given features to predict the solar radio flux
x_new = np.array([[73, 2, 1, 0, 0, 0, 0, 0]])
x_new_scaled = scaler.transform(x_new)
predicted_radio_flux = model.predict(x_new_scaled)
print("Predicted radio flux:", predicted_radio_flux[0,0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step
Predicted radio flux: 114.72667


In [23]:
# Compute 95% heuristic prediction interval
y_pred = predicted_radio_flux[0, 0]
margin = 1.96 * test_mae
lower_CI = y_pred - margin
upper_CI = y_pred + margin

print(f"95% prediction interval: [{lower_CI:.2f}, {upper_CI:.2f}]")

95% prediction interval: [95.22, 134.24]


Using the trained neural network model, **the predicted solar radio flux** for a day when 73 sunspots are observed on the sun and 2 solar flares with C-class energy levels and 1 solar flare with an M-class energy level occur is
approximately **114.73**.
The neural network achieves **a MAE of approximately 9.95 on the test set**, which provides an empirical estimate of the prediction error on unseen data. Because neural networks do not provide statistical standard errors, a statistical
**prediction interval cannot be constructed** as in linear regression model. Instead, the test-set MAE was used to construct **a heuristic prediction interval**, which is roughly **95.22 to 134.24**. I am **moderately confident** in the prediction.